### PR TITLE
:bug: Preallocate free ports before running network tests

### LINKED ISSUE #2502: [flaky] networking.TestDialTLSWithBackoffSuccess
Failed run: https://prow.knative.dev/view/gs/knative-prow/pr-logs/pull/knative_pkg/2500/unit-tests_pkg_main/1519305375889756160
```
=== RUN   TestDialTLSWithBackoffSuccess
2022/04/27 13:21:09 http: TLS handshake error from 127.0.0.1:46468: EOF
2022/04/27 13:21:09 http: TLS handshake error from 127.0.0.1:46472: EOF
2022/04/27 13:21:09 http: TLS handshake error from 127.0.0.1:46470: EOF
    transports_test.go:167: Dial error = timed out dialing after 0.22s
2022/04/27 13:21:09 http: TLS handshake error from 127.0.0.1:46474: write tcp 127.0.0.1:42715->127.0.0.1:46474: use of closed network connection
--- FAIL: TestDialTLSWithBackoffSuccess (0.23s)
```

/assign @cardil

### PR DESCRIPTION
**Pull request by cardil**
# Changes

- :bug: Preallocate free ports before running network tests

/kind bug

Fixes #2502

### CODE REVIEW

**Comment by cardil** [comment]
Forcing a retest to validate robustness a bit.

/test unit-tests_pkg_main

**Comment by dprotaso** [comment]
I still get a consistent failure on Darwin

```
--- FAIL: TestDialWithBackoff (0.48s)
    transports_test.go:86: Unused ports: [63232 63233 63234 63235 63236 63237]
    --- FAIL: TestDialWithBackoff/PlainText (0.00s)
        --- FAIL: TestDialWithBackoff/PlainText/Timeout (0.00s)
            transports_test.go:115: TestDialWithBackoff/PlainText/Timeout port: 63233
            transports_test.go:133: Unexpected success dialing
```

File: network/transports_test.go [hunk 1]
```diff
@@ -208,13 +241,29 @@ func closeOrFail(t testingT, con io.Closer) {
...
+			defer closeOrFail(t, l)
+			p[i] = l.Addr().(*net.TCPAddr).Port
+		}(i)
 	}
-	defer closeOrFail(t, l)
-	return l.Addr().(*net.TCPAddr).Port
+	return &ports{numbers: p}
+}
+
+type ports struct {
```
Comment (dprotaso) [inline network/transports_test.go:259]: nit: https://pkg.go.dev/k8s.io/apimachinery/pkg/util/sets#Int

**Comment by dprotaso** [comment]
Based on this I don't think this test will ever work on darwin
https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/listen.2.html

```
BUGS
     The backlog is currently limited (silently) to 128.
```

**Comment by cardil** [comment]
At @dprotaso:

> The backlog is currently limited (silently) to 128.

Nice catch!

I see possible options:

- rollback timeout test to use non routable IP (only for Darwin)
- skip timeout test on Darwin

Maybe you've opinion on this @evankanderson? It was your idea to use socket with backlog equal 1.

**Comment by evankanderson** [comment]
> Based on this I don't think this test will ever work on darwin https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/listen.2.html
> 
> ```
> BUGS
>      The backlog is currently limited (silently) to 128.
> ```

I think that's an upper bound, though, not a lower bound?

**Review by evankanderson — COMMENTED** [review]
I'm concerned about the strategy of opening a set of TCP ports to get the port numbers, then closing the ports and then hoping to open them again later.

It feels like it would be safer to open the port and then keep it open than to open 6 ports, close them, and then attempt to open them again, particularly since this use of `listen` will probably chose from the high-numbered (ephemeral) port range.

**Comment by evankanderson** [comment]
The changes I used to validate a `listen` backlog of 1 working on darwin/MacOS:

```diff
--- a/network/transports_test.go
+++ b/network/transports_test.go
@@ -26,7 +26,7 @@ import (
        "net"
        "net/http"
        "net/http/httptest"
-       "runtime"
+       //"runtime"
        "strings"
        "syscall"
        "testing"
@@ -120,11 +120,13 @@ func dialTimeout(c dialWithBackoffTestConfig) func(t *testing.T) {
                        t.Fatal("Unable to create listener:", err)
                }
                defer closer()
+               for i := 0; i < 1; i++ {
                c1, err := net.Dial("tcp4", addr.String())
                if err != nil {
-                       t.Fatalf("Unable to connect to server on %s: %s", addr, err)
+                       t.Fatalf("Attempt %d: Unable to connect to server on %s: %s", i, addr, err)
                }
                defer closeOrFail(t, c1)
+                }
 
                // Since the backlog is full, the next request must time out.
                dial := newDialer(ctx, c.tlsConf)
@@ -273,11 +275,11 @@ func newTestErr(msg string, err error) error {
 // net.ListenTCP, so we need to get directly into syscall land.
 func listenOne(t testingT, c dialWithBackoffTestConfig) (func(), *net.TCPAddr, error) {
        //goland:noinspection GoBoolExpressions
-       if runtime.GOOS == "darwin" {
-               // See: https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/listen.2.html
-               t.Skipf("BUG: listen(2) on darwin: The backlog is currently " +
-                       "limited (silently) to 128.")
-       }
+//     if runtime.GOOS == "darwin" {
+//             // See: https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/listen.2.html
+//             t.Skipf("BUG: listen(2) on darwin: The backlog is currently " +
+//                     "limited (silently) to 128.")
+//     }
        fd, err := syscall.Socket(syscall.AF_INET, syscall.SOCK_STREAM, 0)
        if err != nil {
                return nil, nil, newTestErr("Couldn't get socket", err)
@@ -289,7 +291,7 @@ func listenOne(t testingT, c dialWithBackoffTestConfig) (func(), *net.TCPAddr, e
        if err = syscall.Bind(fd, sa); err != nil {
                return nil, nil, newTestErr("Unable to bind", err)
        }
-       if err = syscall.Listen(fd, 0); err != nil {
+       if err = syscall.Listen(fd, 1); err != nil {
                return nil, nil, newTestErr("Unable to Listen", err)
        }
        closer := func() { _ = syscall.Close(fd) }
```

**Comment by dprotaso** [comment]
Closing in favour of https://github.com/knative/pkg/pull/2510

**Pull request closed without merging**